base/compiler/tfuncs.jl

```diff
@@ -1699,6 +1709,87 @@ function _builtin_nothrow(@nospecialize(f), argtypes::Array{Any,1}, @nospecializ
     return false
 end
 
+# known to be always effect-free (in particular nothrow)
+const _PURE_BUILTINS = Any[tuple, svec, ===, typeof, nfields]
+
+# known to be effect-free (but not necessarily nothrow)
+const _EFFECT_FREE_BUILTINS = [
+    fieldtype, apply_type, isa, UnionAll,
+    getfield, arrayref, const_arrayref, isdefined, Core.sizeof,
+    Core.kwfunc, Core.ifelse, Core._typevar, (<:),
+    typeassert, throw, arraysize
+]
+
+const _IDEMPOTENT_BUILTINS = Any[
+    tuple, # tuple is immutable, thus tuples of egal arguments are egal
+    ===,
+    typeof,
+    nfields,
+    fieldtype,
+    apply_type,
+    isa,
+    UnionAll,
+    isdefined,
```

Comment:
This implies all of getfield/arrayref/etc too?

Reply:
`isdefined` has special handling, it should be dropped from this list, but is otherwise harmless.